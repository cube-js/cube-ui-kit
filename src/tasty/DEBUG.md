# Tasty CSS Runtime Debug Utilities

The tasty system provides powerful debug utilities to inspect generated CSS at runtime. These are especially useful for debugging complex responsive layouts and state-based styling.

## Basic Usage

```typescript
import { tastyDebug } from '@cube-dev/ui-kit';

// üéØ Inspect an element by selector
const css = tastyDebug.inspectElement('[data-qa="my-header"]');

// üéØ Get CSS for a specific tasty class
const classCSS = tastyDebug.getCSSForClass('t24');

// üéØ Inspect a DOM element directly
const element = document.querySelector('.my-component');
tastyDebug.inspectDOMElement(element);

// üéØ Get all generated CSS
const allCSS = tastyDebug.getAllCSS();

// üéØ Get a summary of all tasty usage
tastyDebug.getSummary();
```

## Browser Console Usage

In development mode, `tastyDebug` is automatically available on `window`:

```javascript
// Check what's available
tastyDebug.getSummary()

// Inspect your components
tastyDebug.inspectElement('[data-qa="header"]')
tastyDebug.inspectElement('.some-component')

// Check specific classes
tastyDebug.getCSSForClass('t24')

// Find all tasty classes in use
tastyDebug.findAllTastyClasses()
```

## Advanced Inspection

```typescript
// Detailed inspection with breakdown
const inspection = tastyDebug.inspect('[data-qa="complex-grid"]');
console.log('Element:', inspection.element);
console.log('Tasty classes:', inspection.tastyClasses);
console.log('Full CSS:', inspection.css);
console.log('Per-class breakdown:', inspection.breakdown);

// Pretty-print CSS
tastyDebug.logCSS(someCSS, 'My Component CSS');
```

## Example Output

When you run `tastyDebug.inspectElement('[data-qa="header"]')`, you'll see:

```css
.t24:not([data-is-sticky]) {
  position: static; 
  background-color: transparent; 
  z-index: auto; 
  box-shadow: none var(--shadow-color);
}

.t24[data-is-sticky] {
  position: sticky; 
  top: 0; 
  background-color: var(--white-color); 
  z-index: 10; 
  box-shadow: 0 calc(0.5 * var(--gap)) calc(0.5 * var(--gap)) var(--white-color);
}

.t24 {
  display: grid; 
  padding: var(--gap) calc(2 * var(--gap)); 
  height: auto; 
  min-height: calc(6 * var(--gap)); 
  place-items: center stretch; 
  place-content: stretch;
}

@media (min-width: 1200px) {
  .t24:not([data-is-back-button-top]) {
    grid-template-areas: 
      "breadcrumbs breadcrumbs breadcrumbs breadcrumbs breadcrumbs breadcrumbs"
      "back        title       .           subtitle    spacer      extra"
      "back        title       .           subtitle    spacer      extra"
      "content     content     content     content     content     content"
      "footer      footer      footer      footer      footer      footer";
  }
}

/* ...more responsive and state-based CSS... */
```

## Use Cases

- **üêõ Debugging**: Why isn't my grid layout working?
- **üîç Understanding**: What CSS is actually being generated?
- **‚ö° Performance**: How much CSS is being generated?
- **üì± Responsive**: Are my breakpoints working correctly?
- **üéõÔ∏è State Logic**: Are my modifiers applying the right styles?

## Environment Behavior

### Development Mode
The global `window.tastyDebug` is automatically available when `NODE_ENV` is 'development' or not 'production':

```javascript
// Available automatically in browser console
tastyDebug.getSummary()
```

### Production Mode
In production, the global `window.tastyDebug` is not installed automatically. You can either:

1. **Import directly** (recommended):
```typescript
import { tastyDebug } from '@cube-dev/ui-kit';
tastyDebug.inspectElement('[data-qa="my-component"]');
```

2. **Manually install globally**:
```typescript
import { installGlobalDebug } from '@cube-dev/ui-kit';
installGlobalDebug(); // Makes window.tastyDebug available
```

### Manual Installation

You can manually install the global debug utilities at any time:

```typescript
import { installGlobalDebug, tastyDebug } from '@cube-dev/ui-kit';

// Install globally for easy console access
installGlobalDebug();

// Or use directly
tastyDebug.inspectElement('[data-qa="header"]');
```
