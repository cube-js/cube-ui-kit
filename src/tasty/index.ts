export { tasty, Element } from './tasty';
export * from './utils/filterBaseProps';
export * from './utils/colors';
export * from './utils/styles';
export * from './utils/modAttrs';
export * from './utils/responsive';
export * from './utils/renderStyles';
export * from './utils/dotize';
export * from './styles/list';
export * from './providers/BreakpointsProvider';
export * from './utils/mergeStyles';
export * from './utils/warnings';
export * from './utils/getDisplayName';
export type {
  TastyProps,
  GlobalTastyProps,
  AllBasePropsWithMods,
} from './tasty';
export type {
  AllBaseProps,
  BaseProps,
  BaseStyleProps,
  DimensionStyleProps,
  ColorStyleProps,
  OuterStyleProps,
  PositionStyleProps,
  TextStyleProps,
  BlockStyleProps,
  ContainerStyleProps,
  BasePropsWithoutChildren,
  Props,
  FlowStyleProps,
  ShortGridStyles,
  GlobalStyledProps,
  TagNameProps,
  TagName,
} from './types';
export type {
  StylesInterface,
  Styles,
  StylesWithoutSelectors,
  NoType,
  Selector,
  SuffixForSelector,
  NotSelector,
} from './styles/types';
