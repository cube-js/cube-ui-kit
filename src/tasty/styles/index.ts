import { predefine } from './predefined';

const { STYLE_HANDLER_MAP, defineCustomStyle, defineStyleAlias } = predefine();

export { STYLE_HANDLER_MAP, defineCustomStyle, defineStyleAlias };
export * from './createStyle';
