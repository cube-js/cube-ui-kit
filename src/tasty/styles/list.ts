export const BASE_STYLES = [
  'display',
  'font',
  'preset',
  'hide',
  'opacity',
  'whiteSpace',
] as const;

export const POSITION_STYLES = [
  'gridArea',
  'order',
  'gridColumn',
  'gridRow',
  'placeSelf',
  'alignSelf',
  'justifySelf',
  'zIndex',
  'margin',
] as const;

export const BLOCK_STYLES = [
  'reset',
  'padding',
  'paddingInline',
  'paddingBlock',
  'shadow',
  'border',
  'radius',
  'opacity',
  'overflow',
  'styledScrollbar',
  'hide',
  'outline',
  'textAlign',
] as const;

export const COLOR_STYLES = ['color', 'fill'] as const;

export const TEXT_STYLES = [
  'textTransform',
  'fontWeight',
  'fontStyle',
] as const;

export const DIMENSION_STYLES = [
  'width',
  'height',
  'flexBasis',
  'flexGrow',
  'flexShrink',
  'flex',
] as const;

export const FLOW_STYLES = [
  'flow',
  'placeItems',
  'placeContent',
  'alignItems',
  'alignContent',
  'justifyItems',
  'justifyContent',
  'align',
  'justify',
  'gap',
  'columnGap',
  'rowGap',
  'gridColumns',
  'gridRows',
  'gridTemplate',
  'gridAreas',
] as const;

export const CONTAINER_STYLES = [
  ...BASE_STYLES,
  ...COLOR_STYLES,
  ...DIMENSION_STYLES,
  ...POSITION_STYLES,
  ...BLOCK_STYLES,
  ...FLOW_STYLES,
] as const;

export const OUTER_STYLES = [...POSITION_STYLES, ...DIMENSION_STYLES] as const;
