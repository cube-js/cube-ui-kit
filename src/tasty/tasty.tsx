import styledComponents, { createGlobalStyle } from 'styled-components';
import { ComponentType, FC, forwardRef, useContext, useMemo } from 'react';
import { isValidElementType } from 'react-is';

import { BreakpointsContext } from './providers/BreakpointsProvider';
import { modAttrs } from './utils/modAttrs';
import { AllBaseProps, BaseProps, BaseStyleProps, Props } from './types';
import { renderStyles } from './utils/renderStyles';
import { pointsToZones } from './utils/responsive';
import { Styles, StylesInterface } from './styles/types';
import { BASE_STYLES } from './styles/list';
import { ResponsiveStyleValue } from './utils/styles';
import { cacheWrapper } from './utils/cache-wrapper';
import { mergeStyles } from './utils/mergeStyles';
import { getDisplayName } from './utils/getDisplayName';

type StyleList = readonly (keyof {
  [key in keyof StylesInterface]: StylesInterface[key];
})[];

export type PropsWithStyles = {
  styles?: Styles;
} & Omit<Props, 'styles'>;

export type VariantMap = Record<string, Styles>;

export type WithVariant<V extends VariantMap> = {
  variant?: keyof V;
};

export type TastyProps<
  K extends StyleList,
  V extends VariantMap,
  DefaultProps = Props,
> = {
  /** The tag name of the element. */
  as?: string;
  /** Default styles of the element. */
  styles?: Styles;
  /** The list of styles that can be provided by props */
  styleProps?: K;
  element?: BaseProps['element'];
  variants?: V;
} & Partial<Omit<DefaultProps, 'as' | 'styles' | 'styleProps'>> &
  Pick<BaseProps, 'qa' | 'qaVal'> &
  WithVariant<V>;

export interface GlobalTastyProps {
  breakpoints?: number[];
}

export type AllBasePropsWithMods<K extends StyleList> = AllBaseProps & {
  [key in K[number]]?: ResponsiveStyleValue<StylesInterface[key]>;
} & BaseStyleProps;

type TastyComponentPropsWithDefaults<
  Props extends PropsWithStyles,
  DefaultProps extends Partial<Props>,
> = keyof DefaultProps extends never
  ? Props
  : {
      [key in Extract<keyof Props, keyof DefaultProps>]?: Props[key];
    } & {
      [key in keyof Omit<Props, keyof DefaultProps>]: Props[key];
    };

function tasty<K extends StyleList, V extends VariantMap>(
  options: TastyProps<K, V>,
  secondArg?: never,
): ComponentType<Omit<Props, 'variant'> & WithVariant<V>>;
function tasty(selector: string, styles?: Styles);
function tasty<
  Props extends PropsWithStyles,
  DefaultProps extends Partial<Props> = Partial<Props>,
>(
  Component: ComponentType<Props>,
  options?: TastyProps<never, never, Props>,
): ComponentType<TastyComponentPropsWithDefaults<Props, DefaultProps>>;

// Implementation
function tasty<
  K extends StyleList,
  V extends VariantMap,
  C = Record<string, unknown>,
>(Component, options) {
  if (typeof Component === 'string') {
    let selector = Component;
    let styles = options;
    let Element = createGlobalStyle`${(css) => css}`;

    const _StyleDeclarationComponent: FC<GlobalTastyProps> = ({
      breakpoints,
    }) => {
      let contextBreakpoints = useContext(BreakpointsContext);

      breakpoints = breakpoints ?? contextBreakpoints;

      let css = useMemo(
        () =>
          styles
            ? `\n{}${selector}{${renderStyles(
                styles,
                pointsToZones(breakpoints || contextBreakpoints),
              )}}`
            : '',
        [breakpoints.join(',')],
      );

      return <Element css={css} />;
    };

    _StyleDeclarationComponent.displayName = `TastyStyleDeclaration(${Component})`;

    return _StyleDeclarationComponent;
  }

  if (isValidElementType(Component)) {
    let {
      as: extendTag,
      element: extendElement,
      ...defaultProps
    } = options ?? {};

    let propsWithStyles = ['styles'].concat(
      Object.keys(defaultProps).filter((prop) => prop.endsWith('Styles')),
    );

    let _WrappedComponent = forwardRef((props: C, ref) => {
      const { as, element, ...restProps } =
        props as unknown as AllBasePropsWithMods<K>;
      const propsWithStylesValues = propsWithStyles.map((prop) => props[prop]);

      const mergedStylesMap: Styles | undefined = useMemo(() => {
        return propsWithStyles.reduce((map, prop) => {
          if (restProps[prop] != null && defaultProps[prop] != null) {
            map[prop] = mergeStyles(defaultProps[prop], restProps[prop]);
          } else {
            map[prop] = restProps[prop] || defaultProps[prop];
          }

          return map;
        }, {});
      }, propsWithStylesValues);

      return (
        <Component
          ref={ref}
          {...defaultProps}
          {...restProps}
          {...mergedStylesMap}
          as={as ?? extendTag}
          element={element || extendElement}
        />
      );
    });

    _WrappedComponent.displayName = `TastyWrappedComponent(${getDisplayName(
      Component,
      defaultProps.qa ?? extendTag ?? 'Anonymous',
    )})`;

    return _WrappedComponent;
  }

  const tastyOptions = Component as TastyProps<K, never, V>;

  let {
    as: originalAs = 'div',
    element: defaultElement,
    styles: defaultStyles,
    styleProps,
    variants,
    ...defaultProps
  } = tastyOptions;

  let _TastyComponent;

  if (variants) {
    /**
     * Create a tasty component for each variant.
     * Then return a component that receives a `variant` and chooses the right component.
     */
    const variantComponents = Object.keys(variants).reduce((map, variant) => {
      const variantStyles = variants?.[variant];

      map[variant] = tasty({
        as: originalAs,
        styles: mergeStyles(defaultStyles, variantStyles),
        styleProps,
        ...defaultProps,
      });

      return map;
    }, {});

    if (!variantComponents['default']) {
      variantComponents['default'] = tasty({
        as: originalAs,
        styles: defaultStyles,
        styleProps,
        ...defaultProps,
      });
    }

    // eslint-disable-next-line react/display-name
    _TastyComponent = forwardRef(
      (allProps: AllBasePropsWithMods<K> & WithVariant<V>, ref) => {
        const { variant, ...restProps } = allProps;

        const Component = (variantComponents[variant as string] ||
          variantComponents['default']) as ComponentType<
          AllBasePropsWithMods<K>
        >;

        return <Component {...restProps} ref={ref} />;
      },
    );
  } else {
    let Element = styledComponents[originalAs](({ css }) => css);

    /**
     * An additional optimization that allows to avoid rendering styles across various instances
     * of the same element if no custom styles are provided via `styles` prop or direct style props.
     */
    const renderDefaultStyles = cacheWrapper((breakpoints) => {
      return renderStyles(defaultStyles || {}, pointsToZones(breakpoints));
    });

    let {
      qa: defaultQa,
      qaVal: defaultQaVal,
      ...otherDefaultProps
    } = defaultProps ?? {};

    // eslint-disable-next-line react/display-name
    _TastyComponent = forwardRef(
      (allProps: AllBasePropsWithMods<K> & WithVariant<V>, ref) => {
        let {
          as,
          styles,
          variant,
          breakpoints,
          mods,
          element,
          qa,
          qaVal,
          ...otherProps
        } = allProps;

        let propStyles: Styles | null = (
          (styleProps
            ? (styleProps as StyleList).concat(BASE_STYLES)
            : BASE_STYLES) as StyleList
        ).reduce((map, prop) => {
          if (prop in otherProps) {
            map[prop] = otherProps[prop];

            delete otherProps[prop];
          }

          return map;
        }, {});

        if (Object.keys(propStyles).length === 0) {
          propStyles = null;
        }

        if (!styles || (styles && Object.keys(styles).length === 0)) {
          styles = undefined;
        }

        const propStylesCacheKey = JSON.stringify(propStyles);
        const stylesCacheKey = useMemo(() => {
          return JSON.stringify(styles);
        }, [styles]);

        const useDefaultStyles = !propStyles && !styles;

        const styleCacheKey = useMemo(() => {
          return `${styles ? JSON.stringify(styles) : ''}.${
            propStyles ? JSON.stringify(propStyles) : ''
          }`;
        }, [propStylesCacheKey, stylesCacheKey]);

        let allStyles: Styles | undefined = useMemo(
          () =>
            useDefaultStyles
              ? defaultStyles
              : mergeStyles(defaultStyles, styles, propStyles),
          [styleCacheKey],
        );

        let contextBreakpoints = useContext(BreakpointsContext);

        breakpoints = breakpoints ?? contextBreakpoints;

        let renderedStyles = useMemo(() => {
          return useDefaultStyles
            ? renderDefaultStyles(breakpoints as number[])
            : renderStyles(allStyles, pointsToZones(breakpoints as number[]));
        }, [allStyles, breakpoints]);

        if (mods) {
          Object.assign(otherProps, modAttrs(mods));
        }

        return (
          <Element
            as={as ?? originalAs}
            data-element={element || defaultElement}
            data-qa={qa || defaultQa}
            data-qaval={qaVal || defaultQaVal}
            {...otherDefaultProps}
            {...otherProps}
            ref={ref}
            css={renderedStyles}
          />
        );
      },
    );
  }

  _TastyComponent.displayName = `TastyComponent(${
    defaultProps.qa || originalAs
  })`;

  return _TastyComponent;
}

const Element = tasty({});

export { tasty, Element };
