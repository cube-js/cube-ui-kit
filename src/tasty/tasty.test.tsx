import { getByTestId, render } from '@testing-library/react';

import { Button } from '../components/actions';
import { Block } from '../components/Block';

import { tasty } from './tasty';
import { CONTAINER_STYLES } from './styles/list';

describe('tasty() API', () => {
  it('should provide defaults and give ability to override', () => {
    const SButton = tasty(Button, { type: 'primary' });

    const { getByTestId, rerender } = render(
      <SButton qa="button" label="Button" />,
    );
    expect(getByTestId('button').dataset.type).toBe('primary');

    rerender(<SButton type="secondary" qa="button" label="Button" />);
    expect(getByTestId('button').dataset.type).toBe('secondary');
  });

  it('should pass styles from tasty', () => {
    const StyledBlock = tasty(Block, { styles: { color: '#clear.1' } });
    const { container } = render(<StyledBlock />);

    expect(container).toMatchSnapshot();
  });

  it('should support modifiers', () => {
    const StyledBlock = tasty(Block, {
      styles: { color: { '': '#dark', modified: '#purple' } },
    });
    const { container } = render(<StyledBlock mods={{ modified: true }} />);

    expect(container).toMatchSnapshot();
  });

  it('should support kebab-case modifiers', () => {
    const StyledBlock = tasty(Block, {
      styles: { color: { '': '#dark', 'somehow-modified': '#purple' } },
    });
    const { container } = render(
      <StyledBlock mods={{ 'somehow-modified': true }} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should support camelCase modifiers', () => {
    const StyledBlock = tasty(Block, {
      styles: { color: { '': '#dark', somehowModified: '#purple' } },
    });
    const { container } = render(
      <StyledBlock mods={{ somehowModified: true }} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should merge styles', () => {
    const Block = tasty({
      styles: {
        color: { '': '#dark', modified: '#purple' },
        fill: '#white',
      },
    });
    const StyledBlock = tasty(Block, {
      styles: { fill: '#black' },
    });
    const { container } = render(<StyledBlock />);

    expect(container).toMatchSnapshot();
  });

  it('should merge styles in custom prop', () => {
    const Block = tasty({
      inputStyles: {
        color: { '': '#dark', modified: '#purple' },
        fill: '#white',
      },
    });
    const StyledBlock = tasty(Block, {
      inputStyles: { fill: '#black' },
    });
    const { container } = render(<StyledBlock />);

    expect(container).toMatchSnapshot();
  });

  it('should be able to override styles', () => {
    const StyledBlock = tasty(Block, { styles: { color: '#clear.1' } });
    const { container } = render(
      <StyledBlock styles={{ color: '#black.1' }} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should support variants', () => {
    const StyledBlock = tasty({
      styles: { color: '#clear' },
      variants: {
        custom: {
          color: '#black',
        },
      },
    });
    const { container } = render(<StyledBlock variant="custom" />);

    expect(container).toMatchSnapshot();
  });

  it('should fallback to default variant', () => {
    const StyledBlock = tasty({
      styles: { color: '#clear' },
      variants: {
        default: {
          color: '#white',
        },
        custom: {
          color: '#black',
        },
      },
    });
    const { container } = render(<StyledBlock />);

    expect(container).toMatchSnapshot();
  });

  it('should pass qa prop', () => {
    const StyledBlock = tasty({ qa: 'Field' });
    const { container } = render(<StyledBlock />);

    expect(getByTestId(container, 'Field', {})).toBeDefined();
  });

  it('should create responsive styles', () => {
    const StyledBlock = tasty(Block, { styles: { display: ['grid', 'flex'] } });
    const { container } = render(<StyledBlock />);

    expect(container).toMatchSnapshot();
  });

  it('should create element styles', () => {
    const Block = tasty({
      styles: { Element: { color: { '': '#dark', modified: '#purple' } } },
    });
    const { container } = render(<Block />);

    expect(container).toMatchSnapshot();
  });

  it('should merge element styles', () => {
    const Block = tasty({
      styles: {
        Element: {
          color: { '': '#dark', modified: '#purple' },
          fill: '#white',
        },
      },
    });
    const StyledBlock = tasty(Block, {
      styles: { Element: { fill: '#black' } },
    });
    const { container } = render(<StyledBlock />);

    expect(container).toMatchSnapshot();
  });

  it('should define style props', () => {
    const Block = tasty({
      styles: {
        border: '2bw',
      },
      styleProps: CONTAINER_STYLES,
    });

    const { container } = render(<Block border={true} />);

    expect(container).toMatchSnapshot();
  });

  it('should allow multiple wrapping', () => {
    const Block = tasty({
      styles: {
        position: 'relative',
        padding: '1x top',
        border: true,
      },
      styleProps: CONTAINER_STYLES,
    });
    const SecondBlock = tasty(Block, {
      styles: {
        border: false,
      },
    });
    const ThirdBlock = tasty(SecondBlock, {
      styles: {
        position: 'static',
        padding: '2x top',
        color: '#white',
        border: '#black',
      },
    });

    const { container } = render(<ThirdBlock display="flex" />);

    expect(container).toMatchSnapshot();
  });

  it.only('should allow nested modifiers', () => {
    const StyledBlock = tasty(Block, {
      styles: { color: { '': '#clear', ':not(:first-child)': '#black' } },
    });
    const { container } = render(<StyledBlock />);

    expect(container).toMatchSnapshot();
  });
});
