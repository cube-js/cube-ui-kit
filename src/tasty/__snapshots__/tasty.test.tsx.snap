// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tasty() API should allow multiple wrapping 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0 {
  position: static;
}

.c0.c0 {
  padding-top: calc(2 * var(--gap));
}

.c0.c0 {
  border: var(--border-width) solid var(--black-color);
}

.c0.c0 {
  color: var(--white-color,rgba(0,0,0,1));
  --current-color: var(--white-color,white);
  --current-color-rgb: var(--white-color-rgb);
}

.c0.c0 {
  display: flex;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should allow nested modifiers 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0:first-child {
  color: var(--clear-color);
  --current-color: var(--clear-color,clear);
  --current-color-rgb: var(--clear-color-rgb);
}

.c0.c0:not(:first-child) {
  color: var(--black-color);
  --current-color: var(--black-color,black);
  --current-color-rgb: var(--black-color-rgb);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should be able to override styles 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0 {
  color: rgba(var(--black-color-rgb),0.1);
  --current-color: var(--black-color,black);
  --current-color-rgb: var(--black-color-rgb);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should create element styles 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0 [data-element="Element"] {
  outline: none;
}

.c0 [data-element="Element"][hidden] {
  display: none !important;
}

.c0.c0:not([data-is-modified]) [data-element="Element"] {
  color: var(--dark-color);
  --current-color: var(--dark-color,dark);
  --current-color-rgb: var(--dark-color-rgb);
}

.c0.c0[data-is-modified] [data-element="Element"] {
  color: var(--purple-color,rgba(0,0,0,1));
  --current-color: var(--purple-color,purple);
  --current-color-rgb: var(--purple-color-rgb);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should create responsive styles 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

@media (min-width:980px) {
  .c0.c0 {
    display: grid;
  }
}

@media (max-width:979px) {
  .c0.c0 {
    display: flex;
  }
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should define style props 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0 {
  border: var(--border-width) solid var(--border-color);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should fallback to default variant 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0 {
  color: var(--white-color,rgba(0,0,0,1));
  --current-color: var(--white-color,white);
  --current-color-rgb: var(--white-color-rgb);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should merge element styles 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0 [data-element="Element"] {
  outline: none;
}

.c0 [data-element="Element"][hidden] {
  display: none !important;
}

.c0.c0:not([data-is-modified]) [data-element="Element"] {
  color: var(--dark-color);
  --current-color: var(--dark-color,dark);
  --current-color-rgb: var(--dark-color-rgb);
}

.c0.c0[data-is-modified] [data-element="Element"] {
  color: var(--purple-color,rgba(0,0,0,1));
  --current-color: var(--purple-color,purple);
  --current-color-rgb: var(--purple-color-rgb);
}

.c0.c0 [data-element="Element"] {
  background-color: var(--black-color);
}

.c0.c0 [data-element="Element"]>* {
  --context-fill-color: var(--black-color);
  --context-fill-color-rgb: var(--black-color-rgb);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should merge styles 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0:not([data-is-modified]) {
  color: var(--dark-color);
  --current-color: var(--dark-color,dark);
  --current-color-rgb: var(--dark-color-rgb);
}

.c0.c0[data-is-modified] {
  color: var(--purple-color,rgba(0,0,0,1));
  --current-color: var(--purple-color,purple);
  --current-color-rgb: var(--purple-color-rgb);
}

.c0.c0 {
  background-color: var(--black-color);
}

.c0.c0>* {
  --context-fill-color: var(--black-color);
  --context-fill-color-rgb: var(--black-color-rgb);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should merge styles in custom prop 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should pass styles from tasty 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0 {
  color: rgba(var(--clear-color-rgb),0.1);
  --current-color: var(--clear-color,clear);
  --current-color-rgb: var(--clear-color-rgb);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`tasty() API should support camelCase modifiers 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0:not([data-is-somehow-modified]) {
  color: var(--dark-color);
  --current-color: var(--dark-color,dark);
  --current-color-rgb: var(--dark-color-rgb);
}

.c0.c0[data-is-somehow-modified] {
  color: var(--purple-color,rgba(0,0,0,1));
  --current-color: var(--purple-color,purple);
  --current-color-rgb: var(--purple-color-rgb);
}

<div>
  <div
    class="c0"
    data-is-somehow-modified=""
  />
</div>
`;

exports[`tasty() API should support kebab-case modifiers 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0:not([data-is-somehow-modified]) {
  color: var(--dark-color);
  --current-color: var(--dark-color,dark);
  --current-color-rgb: var(--dark-color-rgb);
}

.c0.c0[data-is-somehow-modified] {
  color: var(--purple-color,rgba(0,0,0,1));
  --current-color: var(--purple-color,purple);
  --current-color-rgb: var(--purple-color-rgb);
}

<div>
  <div
    class="c0"
    data-is-somehow-modified=""
  />
</div>
`;

exports[`tasty() API should support modifiers 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0:not([data-is-modified]) {
  color: var(--dark-color);
  --current-color: var(--dark-color,dark);
  --current-color-rgb: var(--dark-color-rgb);
}

.c0.c0[data-is-modified] {
  color: var(--purple-color,rgba(0,0,0,1));
  --current-color: var(--purple-color,purple);
  --current-color-rgb: var(--purple-color-rgb);
}

<div>
  <div
    class="c0"
    data-is-modified=""
  />
</div>
`;

exports[`tasty() API should support variants 1`] = `
.c0 {
  outline: none;
}

.c0[hidden] {
  display: none !important;
}

.c0.c0 {
  color: var(--black-color);
  --current-color: var(--black-color,black);
  --current-color-rgb: var(--black-color-rgb);
}

<div>
  <div
    class="c0"
  />
</div>
`;
